File: py/clientAis/communications/ServerCommander.py
import time
from typing import List, Set, Union

import numpy as np

from clientAis.communications.JsonHandler import JsonHandler
from clientAis.communications.MessageSerializer import MessageSerializer
from clientAis.networking.CommunicationHandler import CommunicationHandler

class ServerCommander:
    def __init__(self, serverCommunicator: CommunicationHandler, messageSerializer: MessageSerializer):
        self.messageSerializer = messageSerializer
        self.serverCommunicator = serverCommunicator

    def moveUnitsToPosition(self, singuityIds: Union[List[str], Set[str]], position: np.ndarray):
        self.serverCommunicator.send(
            JsonHandler.dump(
                {
                    "type": "move-units-to-position",
                    "data": {"m": singuityIds, "d": {"x": position[0], "y": position[1]}},
                    "timestamp": int(time.time() * 1000)
                }
            ) + "\\n"
        )

    def moveUnitsToSpawner(self, singuityIds: Union[List[str], Set[str]], spawnerId: str):
        self.serverCommunicator.send(
            JsonHandler.dump(
                {
                    "type": "move-units-to-spawner",
                    "data": {"m": singuityIds, "s": spawnerId},
                    "timestamp": int(time.time() * 1000)
                }
            ) + "\\n"
        )

    def setSpawnersRally(self, spawnersId: Union[List[str], Set[str]], mobileDestination: tuple):
        if mobileDestination[0] == "point":
            mobileDestinationJson = {"x": mobileDestination[1][0], "y": mobileDestination[1][1]}
        elif mobileDestination[1] == "allegedSpawner":
            allegenceJson = "n" if mobileDestination[1]["allegence"] is None else {
                "i": mobileDestination[1]["allegence"]["isClaimed"],
                "h": mobileDestination[1]["allegence"]["healthPoints"],
                "p": mobileDestination[1]["allegence"]["playerId"]
            }

            mobileDestinationJson = {"s": {"i": mobileDestination[1]["id"], "a": allegenceJson}}
        else:  # mobileDestination[1] == "spawner"
            mobileDestinationJson = {"u": mobileDestination[1]}

        self.serverCommunicator.send(
            JsonHandler.dump(
                {
                    "type": "set-spawners-rally",
                    "data": {"s": spawnersId, "d": mobileDestinationJson},
                    "timestamp": int(time.time() * 1000)
                }
            ) + "\\n"
        )

File: jva/src/clientAis/communications/ServerCommander.java
package clientAis.communications;

import clientAis.communications.game_data.MobileDestination;
import clientAis.networking.CommunicationHandler;
import net.minidev.json.JSONArray;
import net.minidev.json.JSONObject;
import utils.math.vector.Vector2;

import java.util.Set;

public class ServerCommander {

    private final CommunicationHandler serverCommunicator;
    private final MessageSerializer messageSerializer;

    public ServerCommander(CommunicationHandler serverCommunicator, MessageSerializer messageSerializer) {
        this.serverCommunicator = serverCommunicator;
        this.messageSerializer = messageSerializer;
    }

    public void moveUnitsToPosition(Set<String> singuityIds, Vector2 position) {
        final JSONObject jsonPacket = new JSONObject();
        jsonPacket.put("type", "move-units-to-position");
        final JSONObject jsonData = new JSONObject();
        final JSONArray singuityIdsJson = new JSONArray();
        singuityIdsJson.addAll(singuityIds);
        jsonData.put("m", singuityIdsJson);
        final JSONObject jsonPosition = new JSONObject();
        jsonPosition.put("x", position.x);
        jsonPosition.put("y", position.y);
        jsonData.put("d", jsonPosition);
        jsonPacket.put("data", jsonData);
        jsonPacket.put("timestamp", System.currentTimeMillis());

        serverCommunicator.send(jsonPacket + "\\n");
    }

    public void moveUnitsToSpawner(Set<String> singuityIds, String spawnerId) {
        final JSONObject jsonPacket = new JSONObject();
        jsonPacket.put("type", "move-units-to-spawner");
        final JSONObject jsonData = new JSONObject();
        final JSONArray singuityIdsJson = new JSONArray();
        singuityIdsJson.addAll(singuityIds);
        jsonData.put("m", singuityIdsJson);
        jsonData.put("s", spawnerId);
        jsonPacket.put("data", jsonData);
        jsonPacket.put("timestamp", System.currentTimeMillis());

        serverCommunicator.send(jsonPacket + "\\n");
    }

    public void setSpawnersRally(Set<String> spawnersId, MobileDestination mobileDestination) {
        final JSONObject jsonPacket = new JSONObject();
        jsonPacket.put("type", "set-spawners-rally");
        final JSONObject jsonData = new JSONObject();
        final JSONArray spawnersIdJson = new JSONArray();
        spawnersIdJson.addAll(spawnersId);
        jsonData.put("s", spawnersIdJson);
        jsonData.put("d", mobileDestination.serializeToJson());
        jsonPacket.put("data", jsonData);
        jsonPacket.put("timestamp", System.currentTimeMillis());

        serverCommunicator.send(jsonPacket + "\\n");
    }
}
